File: problem_012.c
// -*- compile-command: "gcc -std=c99 -o problem_012 problem_012.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include "ceuler.h"

int findDivisors(int input, int *primes, int nprimes)
{
    int index, count=1;
    int primeCount[nprimes];
    for (index=0; index<nprimes; index++)
        primeCount[index] = 0;

    // Prime factorization
    while (input != 1)
        for (index=0; index<nprimes; index++)
            if (input % primes[index] == 0)
            {
                input = input / primes[index];
                primeCount[index]++;
            }

    // Convert prime factorization to divisors
    for (index=0; index<nprimes; index++)
        if (primeCount[index] > 0)
            count *= (primeCount[index]+1);

    return count;
}

int main()
{
    printf("Project Euler - Problem 12:\n"
           "What is the value of the first triangle number to have over five hundred divisors?\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    int index=0, triangle=0, divisors=0;
    int nprimes = 1500;
    int primes[nprimes];
    primeSieve(primes, nprimes);

    // Calculate Triangles
    index = 0;
    while(index<1000000)
    {
        index++;
        triangle += index;

        // Skip numbers not divisible by 3 or 5
        if ( (triangle % 3 != 0) || (triangle % 5 != 0) )
            continue;
        divisors = findDivisors(triangle, primes, nprimes);

        if (divisors > 500)
            break;
    }

    printf("Triangle Value: %d, Divisors: %d\n", triangle, divisors);
    printElapsedTime(start);

    return 0;
}

File: problem_012.rs
// -*- compile-command: "rustc -o problem_012_rs problem_012.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn is_prime(n: u64) -> bool {
    if n < 2 {
        return false;
    }
    if n == 2 {
        return true;
    }
    if n % 2 == 0 {
        return false;
    }

    let limit = (n as f64).sqrt() as u64;
    for i in 3..limit+1 {
        if n % i == 0 {
            return false;
        }
    }
    return true;
}

fn divisors(mut n: u64, p: &Vec<u64>) -> u64 {
    let mut count = 1;
    for i in 0..p.len() {
        let mut t = 1;
        while n % p[i] == 0 {
            t += 1;
            n /= p[i];
        }
        count *= t;
        if n == 1 {
            break;
        }
    }
    return count;
}


fn main() {
    println!("Project Euler - Problem 12");
    println!("What is the value of the first triangle number to have over five hundred divisors?\n");

    // Find primes!
    let mut primes = Vec::new();
    for n in 2..5000 {
        if is_prime(n) {
            primes.push(n)
        }
    }

    // Find triangles
    let mut tri: u64 = 1;
    for n in 2..1000000 {
        tri += n;
        if divisors(tri, &primes) > 500 {
            break;
        }
    }


    println!("{}", tri);
}
